shell/common/gin_helper/dictionary.h

```diff
@@ -109,6 +110,36 @@ class Dictionary : public gin::Dictionary {
         .ToChecked();
   }
 
+  template <typename K, typename V>
+  bool SetGetter(const K& key, const V& val) {
+    AccessorValue<V> acc_value;
+    acc_value.Value = val;
+
+    v8::Local<v8::Value> v8_value_accessor;
+    if (!gin::TryConvertToV8(isolate(), acc_value, &v8_value_accessor))
+      return false;
+
+    auto context = isolate()->GetCurrentContext();
+
+    return GetHandle()
+        ->SetAccessor(
+            context, gin::StringToV8(isolate(), key),
+            [](v8::Local<v8::Name> property_name,
+               const v8::PropertyCallbackInfo<v8::Value>& info) {
+              AccessorValue<V> acc_value;
+              if (!gin::ConvertFromV8(info.GetIsolate(), info.Data(),
+                                      &acc_value))
+                return;
+
+              V val = acc_value.Value;
+              v8::Local<v8::Value> v8_value;
+              if (gin::TryConvertToV8(info.GetIsolate(), val, &v8_value))
+                info.GetReturnValue().Set(v8_value);
+            },
+            NULL, v8_value_accessor)
+        .ToChecked();
+  }
```

Comment:
I think `AccessorValue` here is a little obtuse. Perhaps we could take a getter method here as an argument instead? e.g.

```cpp
  template <typename K, typename V>
  bool SetGetter(const K& key, base::RepeatingCallback<V()> getter) {
    return GetHandle()
        ->SetAccessor(
            isolate()->GetCurrentContext(),
            gin::StringToV8(isolate(), key),
            [](v8::Local<v8::Name> property_name,
               const v8::PropertyCallbackInfo<v8::Value>& info) {
              V val = getter.Run();
              v8::Local<v8::Value> v8_value;
              if (gin::TryConvertToV8(info.GetIsolate(), val, &v8_value))
                info.GetReturnValue().Set(v8_value);
            },
            NULL, v8_value_accessor)
        .ToChecked();
  }
```

and then bind the RFH details into the callback, e.g.

```cpp
dict.SetGetter("frame", base::BindRepeating(
  [](int process_id, int routing_id) {
    return content::RenderFrameHost::FromID(process_id, routing_id);
  }, rfh->GetProcess()->GetID(), rfh->GetRoutingID()));
```

Reply:
This is similar to what I had in mind when I initially worked on this. However, I don't believe the `SetAccessor()` API is designed in mind for supporting lambdas with capturing.

https://source.chromium.org/chromium/chromium/src/+/main:v8/src/api/api.cc;l=1584-1591;drc=6ff1129ca37e3e618db1b594452da793331a2f28
```
void ObjectTemplate::SetAccessor(v8::Local<String> name,
                                 AccessorGetterCallback getter,
                                 AccessorSetterCallback setter,
                                 v8::Local<Value> data, AccessControl settings,
                                 PropertyAttribute attribute,
                                 v8::Local<AccessorSignature> signature,
                                 SideEffectType getter_side_effect_type,
                                 SideEffectType setter_side_effect_type)
```
```
using AccessorNameGetterCallback =
    void (*)(Local<Name> property, const PropertyCallbackInfo<Value>& info);
```

The type defined by `AccessorNameGetterCallback` will not allow data to be captured by the lambda afaik. Capturing would be required for the following line in your snippet:
```
V val = getter.Run();
```

This limited my implementation to rely on the `v8::Local<Value> data` parameter for storing data to later be read upon property access. Thus I needed gin converters to save the data in v8 and later convert back to the generic value.

----

One additional concern would be the code duplication involved with the callback approach. If we want to lazily resolve RFH with property getters elsewhere, we'd need to copy and paste the lambda for each.

Reply:
Ah, you're right... I suppose that makes sense from a perspective of knowing how/when to GC the associated data.

Come to think of it, this might be a good fit for `gin::ObjectTemplateBuilder`? There's a method called `SetLazyDataProperty` that looks quite similar to this.

Reply:
er, sorry, I was thinking of `gin::DataObjectBuilder`, which doesn't have a similar method. Rats.

Reply:
I still think this is quite a lot of template machinery for a single-use feature, but I'm going to acquiesce for now. Maybe we'll end up finding more uses for it :)